File: Algorithms/Binary Search/Problem Solved/33_Search_in_Rotated_Sorted_Array.py
# https://leetcode.com/problems/search-in-rotated-sorted-array/description/

class Solution:
    def search(self, nums: list[int], target: int) -> int:

        # * variables for binary search
        start, end = 0, len(nums) - 1  # multiple assignment statement

        while start <= end:
            mid = start + (end - start) // 2

            if nums[mid] == target:
                return mid

            # ^ check for left sorted portion
            elif nums[mid] > nums[start]:
                if nums[mid - 1] >= target >= nums[start]:
                    end = mid - 1

                # ^ otherwise go to right sorted portion
                else:
                    start = mid + 1

            # ^ check in the right sorted portion
            else:
                if nums[end] >= target >= nums[mid + 1]:
                    start = mid + 1
                else:
                    end = mid - 1

        return -1


ans1 = Solution()
# print(ans1.search([4, 5, 6, 7, 0, 1, 2], 0))
print(ans1.search([3, 1], 1))

File: Algorithms/Binary Search/Problem Solved/33_Search_in_Rotated_Sorted_Array.cpp
// https://leetcode.com/problems/search-in-rotated-sorted-array/

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int search(vector<int> &arr, int target)
    {
        //* binary search variables
        int start = 0;
        int end = arr.size() - 1;
        int mid = start + (end - start) / 2;

        while (start <= end)
        {
            //* check if mid index has the target element
            if (arr[mid] == target)
                return mid;

            //^ mid divides the array into two portions
            // from middle where we should check for sorted portion

            //* check whether left portion is sorted
            else if (arr[mid] > arr[start])
            {
                // * check whether target is on the left sorted portion
                if (target >= arr[start] && target < arr[mid])
                    end = mid - 1;
                else
                    start = mid + 1;
            }

            else
            {
                if (target <= arr[end] && target > arr[mid])
                    start = mid + 1;
                else
                    end = mid - 1;
            }

            mid = start + (end - start) / 2;
        }

        return -1;
    }
};

int main()
{
    Solution op;
    vector<int> arr{4, 5, 6, 7, 0, 1, 2};
    cout << "Position is: " << op.search(arr, 0) << endl;

    return 0;
}